paper_framework_sources/p2079_system_execution_context.md

```diff
@@ -178,14 +181,163 @@ public:
+We already know that full platform customisations are possible.
+This approach would delegate the decision of how to allow Intel to replace the platform context with TBB up to the platform implementor.
+It would rely on an agreement between the system vendor and the runtime vendor.
+
+The authors do not have a recommendation, only a wish to see customisation available.
+We should decide how best to achieve it within the standard.
+Assuming we delegate customisation to the platform implementor, what wording would be appropriate for the specification, if any?
+
+## Need for the system_context class
+Our goal is to expose a global shared context to avoid oversubscription of threads in the system and to efficiently share a system thread pool.
+Underneath the `system_context` there is a singleton of some sort, potentially owned by the OS.
+
+The question is how we expose the singleton.
+We have a few obvious options:
+
+ - Explicit context objects, as we've described in R2 and R3 of this paper, where a `system_context` is constructed as any other context might be, and refers to a singleton underneath.
+ - A global `get_system_context()` function that obtains a `system_context` object, or a reference to one, representing the singleton explicitly.
+ - A global `get_system_scheduler()` function that obtains a scheduler from some singleton system context, but does not explicitly expose the context.
+
+The `get_system_context()` function returning by value adds little, it's equivalent to direct construction.
+`get_system_context()` returning by reference and `get_system_scheduler()` have a different lifetime semantic from directly constructed `system_context`.
+
+The main reason for having an explicit by-value context is that we can reason about lifetimes.
+If we only have schedulers, from `get_system_context().get_scheduler()` or from `get_system_scheduler()` then we have to think about how they affect the context lifetime.
+We might want to reference count the context, to ensure it outlives the schedulers, but this adds cost to each scheduler use, and to any downstream sender produced from the scheduler as well that is logically dependent on the scheduler.
+We could alternatively not reference count and assume the context outlives everything in the system, but that leads quickly to shutdown order questions and potential surprises.
+
+By making the context explicit we require users to drain their work before they drain the context.
+In debug builds, at least, we can also add reference counting so that descruction of the context before work completes reports a clear error to ensure that people clean up.
+That is harder to do if the context is destroyed at some point after main completes.
+This lifetime question also applies to construction: we can lazily construct a thread pool before we first use a scheduler to it.
+
+For this reason, and consistency with other discussions about structured concurrency, we opt for an explicit context object here.
```

Comment:
I thought a primary goal of system_context context is to provide a default context, even for the synchronous std algorithms, i.e., when no scheduler is explicitly passed.  Would developers need to construct and hold on to a system context then? Where would they do that, and how would they know that they should do that? So I'm not sure that an explicit context makes sense for those default scenarios.

Reply:
Yes, that's a valid question. There are two different goals. One goal is to make available some standard context, that people can rely on and that can make down to the system thread pool. A second would be to make available some singleton they can grab from anywhere in their code.

I am not particularly keen on the second version of this. We find it leads to lifetime issues in the codebase. It is used, though. One way to deal with that is we simply build library facilities on top of the context. So the schedulerless algorithms can themselves share a singleton in the implementation, and we document them as using the system_context, so we know they are going to share threads. They have to have their own systme_context variable that outlives calls to the algorithms, it's just we don't specify a singleton in the standard.

Reply:
I am not sure I agree with the position that algorithms should have it's own copy of system_context. To be clear I do not object to leave it as is for now but it seems it need deeper thinking about that and from my point of view it would also depend on implementation experience.